### ISSUE TITLE
implement proper `disjoint` for rectangles using their area of overlap

### ISSUE THREAD

**Issue by hypotext** [penrose/penrose#584]
**Is your feature request related to a problem? Please describe.**
Currently, the implementation of `disjoint` uses only 1D overlap between intervals to push rectangles apart. This means the optimizer can only move things in an axis-aligned fashion, resulting in local minima.

**Describe the solution you'd like**
Push rectangles apart according to the amount of overlap between them, as described by @keenancrane in Slack here: https://pen-rose.slack.com/archives/CDN2T1Z5J/p1617913422001300

**Describe alternatives you've considered**
N/A

**Additional context**
Maybe relevant to #582

**Comment by joshpoll** [comment]
copy-pasting overlap code from slack
```
// area A of overlap for two rectangles,
// assuming each rectangle is centered at (xi,yi),
// with width wi and height hi
l = max( x1 - w1/2., x2 - w2/2. ); // leftmost point in either rectangle
r = min( x1 + w1/2., x2 + w2/2. ); // rightmost ...
b = max( y1 - h1/2., y2 - h2/2. ); // bottommost ...
t = min( y1 + h1/2., y2 + h2/2. ); // topmost ...
w = max( 0., r - l ); // width of rectangle (zero if no overlap)
h = max( 0., t - b ); // height ...
A = w h // area (zero if no overlap)
```
[OverlappingRectangles.pdf](https://github.com/penrose/penrose/files/6619729/OverlappingRectangles.pdf)

This will be straightforward to implement once #582 is merged.

**Comment by joshpoll** [comment]
Adding the overlaps (i.e. `w + h`) seems to work better in practice than multiplying them. This is because the product can get rounded to `0` for adjacent boxes. The screenshot below is a very common local minimum on the monoidal category example with multiplication.

<img width="222" alt="Screen Shot 2021-06-10 at 2 33 07 PM" src="https://user-images.githubusercontent.com/21694516/121578707-f5551b80-c9f8-11eb-8fe4-30194bebbc80.png">

**Comment by joshpoll** [comment]
Thought about this more, more. I think what is going on is that using `add` kicks the diagram out of the local minimum, but like Keenan said in the meeting that's basically a hack. The other interpretation I _thought_ was going on is that `add` was or-ing together the overlap costs, but that (i) doesn't make sense because overlapping in one dimension is fine and (ii) was not even happening since there is a special condition for disjoint boxes.

**Comment by joshpoll** [comment]
In reference to #542.

The OG code produces diagram output like this. Neither `mul` nor `add` produces this. My best guess is the original uses `min`, which (i) causes a gradient in just one direction and (ii) is very small when one of the overlaps is small.

![114254809-b9152880-997f-11eb-823a-50f9ca4d4aec](https://user-images.githubusercontent.com/21694516/121595641-78cc3800-ca0c-11eb-9756-fde1ac19637c.png)

`add` more consistently produces diagrams like the following. I think this may be because `add`ing the overlaps yields a higher energy than `mul`ing them when the overlap is small and so this energy kicks the diagram out of bad local minima. This is basically an "accident," because the semantics of `add`ing them is not correct.

<img width="506" alt="Screen Shot 2021-06-10 at 5 00 44 PM" src="https://user-images.githubusercontent.com/21694516/121596461-70283180-ca0d-11eb-9fb4-6d7df3b8a81d.png">

**Comment by hypotext** [comment]
I wouldn't take the monoidal category diagram super seriously as there are a lot of miscellaneous constraints/objectives there for other domain reasons (see the `opt` tab), that make the problem harder to satisfy than the "pure" disjoints (the first two examples).  Basically I think what Keenan suggested is the right thing to do, unless you can point to diagrams in the first two examples that are clearly worse!

**Comment by joshpoll** [comment]
Closing as this was implemented/fixed in #582.
